src/managers/XCursorManager.cpp

```diff
@@ -0,0 +1,447 @@
+        defaultCursor = hyprCursor;
+        return;
+    }
+
+    for (auto const& shape : CURSOR_SHAPE_NAMES) {
+        auto legacyName = getLegacyShapeName(shape);
+        if (legacyName.empty())
+            continue;
+
+        auto it = std::find_if(cursors.begin(), cursors.end(), [&legacyName](auto const& c) { return c->shape == legacyName; });
+
+        if (it == cursors.end()) {
+            Debug::log(LOG, "XCursor failed to find a legacy shape with name {}, skipping", legacyName);
+            continue;
+        }
+
+        auto cursor    = makeShared<SXCursors>();
+        cursor->images = it->get()->images;
+        cursor->shape  = shape;
+
+        cursors.emplace_back(cursor);
+    }
+}
+
+SP<SXCursors> CXCursorManager::getShape(std::string const& shape, int size) {
+    // monitor scaling changed etc, so reload theme with new size.
+    if (size != lastLoadSize)
+        loadTheme(themeName, size);
+
+    // try to get an icon we know if we have one
+    for (auto const& c : cursors) {
+        if (c->shape != shape)
+            continue;
+
+        return c;
+    }
+
+    Debug::log(WARN, "XCursor couldn't find shape {} , using default cursor instead", shape);
+    return defaultCursor;
+}
+
+SP<SXCursors> CXCursorManager::createCursor(std::string const& shape, XcursorImages* xImages) {
+    auto xcursor = makeShared<SXCursors>();
+
+    for (int i = 0; i < xImages->nimage; i++) {
+        auto          xImage = xImages->images[i];
+        SXCursorImage image;
+        image.size    = {(int)xImage->width, (int)xImage->height};
+        image.hotspot = {(int)xImage->xhot, (int)xImage->yhot};
+        image.pixels.resize(xImage->width * xImage->height);
+        std::memcpy(image.pixels.data(), xImage->pixels, xImage->width * xImage->height * sizeof(uint32_t));
+        image.delay = xImage->delay;
+
+        xcursor->images.emplace_back(image);
+    }
+
+    xcursor->shape = shape;
+
+    return xcursor;
+}
+
+std::vector<std::string> CXCursorManager::themePaths(std::string const& theme) {
+    auto const*              path = XcursorLibraryPath();
+    std::vector<std::string> paths;
+
+    auto                     expandTilde = [](std::string const& path) {
+        if (!path.empty() && path[0] == '~') {
+            const char* home = std::getenv("HOME");
+            if (home)
+                return std::string(home) + path.substr(1);
+        }
+        return path;
+    };
+
+    if (path) {
+        std::stringstream ss(path);
+        std::string       item;
+
+        while (std::getline(ss, item, ':')) {
+            auto p = expandTilde(item + "/" + theme + "/cursors");
+
+            if (std::filesystem::exists(p) && std::filesystem::is_directory(p))
+                paths.push_back(p);
+        }
+    }
+
+    return paths;
+}
+
+std::string CXCursorManager::getLegacyShapeName(std::string const& shape) {
+    if (shape == "invalid")
```

Comment:
wouldnt a hashmap be faster

Reply:
doubt it, after the compiler optimize this its gonna matter much at all